File: leetcode/#14.py
# question --> https://leetcode.com/problems/longest-common-prefix/

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        iterations = len(min(strs))
        ans = min(strs)
        
        if len(strs) == 1: 
            return strs[0]
        
        index = 0
        for index in range(iterations):
            temp = ans[index]
            for i in range(len(strs)):
                if strs[i][index] != temp: 
                    return ans[:index] 
        
        return ans[:index + 1]
        
        # TC: O(N^2); SC: O(N) 
        
# end 

File: leetcode/#14.js
// question --> https://leetcode.com/problems/longest-common-prefix/

/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function(strs) {
    let ans = strs.reduce((a, b) => a.length <= b.length ? a : b);
    let iterations = ans.length; 
    
    let index = 0; 
    
    for (index; index < iterations; index++) {
        let temp = ans[index];
        for (let i = 0; i < strs.length; i++) {
            if (strs[i][index] != temp) {
                if (index == 0) {
                    return ''
                }
                return ans.slice(0, index);
            } 
        }
    }
    return ans.slice(0, index + 1);
};

// TC: O(N^2); SC: O(N)

// end 